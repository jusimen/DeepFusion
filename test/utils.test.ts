import {
  countOccurrences,
  getInMultiple,
  getInAll,
  getNotInMultiple,
  getNotInAll,
  getObjectKeys,
  isObject,
  isPropDescriptor
} from '../src/utils';

const arr1 = ['a', 'b', 'c'];
const arr2 = ['b', 'c', 'd'];
const arr3 = ['c', 'd', 'e'];
const testObj = Object.create(
  // Prototype (standard object)
  Object.create(null, {
    enumerableTrue: {
      configurable: true,
      enumerable: true,
      value: true,
      writable: true
    },
    enumerableFalse: {
      configurable: true,
      enumerable: false,
      value: false,
      writable: true
    }
  }),
  // Properties (descriptors object)
  {
    firstName: {
      enumerable: true,
      value: 'John',
      writable: true
    },
    lastName: {
      enumerable: true,
      value: 'Smith',
      writable: true
    },
    fullName: {
      enumerable: false,
      get() {
        return `${this.firstName} ${this.lastName}`;
      },
      set(val) {
        const names = val.replace(/\s+/g, ' ').trim().split(' ');

        this.firstName = names[0] || '';
        this.lastName = names[1] || '';
      }
    }
  }
);

describe('countOccurrences', () => {
  test('generates map', () => {
    const result = countOccurrences(arr1, arr2, arr3);

    expect(result).toMatchSnapshot();
  });
});

describe('getInMultiple', () => {
  test('gets common values in two arrays', () => {
    const result = getInMultiple(arr1, arr2);

    expect(result).toMatchSnapshot();
  });

  test('gets common values found in multiple arrays', () => {
    const result = getInMultiple(arr1, arr2, arr3);

    expect(result).toMatchSnapshot();
  });
});

describe('getInAll', () => {
  test('gets universal values in two arrays', () => {
    const result = getInAll(arr1, arr2);

    expect(result).toMatchSnapshot();
  });

  test('gets universal values in multiple arrays', () => {
    const result = getInAll(arr1, arr2, arr3);

    expect(result).toMatchSnapshot();
  });
});

describe('getNotInMultiple', () => {
  test('gets values found in one of two arrays', () => {
    const result = getNotInMultiple(arr1, arr2);

    expect(result).toMatchSnapshot();
  });

  test('gets values found in one of multiple arrays', () => {
    const result = getNotInMultiple(arr1, arr2, arr3);

    expect(result).toMatchSnapshot();
  });
});

describe('getNotInAll', () => {
  test('gets values not found in two arrays', () => {
    const result = getNotInAll(arr1, arr2);

    expect(result).toMatchSnapshot();
  });

  test('gets values not found in three arrays', () => {
    const result = getNotInAll(arr1, arr2, arr3);

    expect(result).toMatchSnapshot();
  });
});

describe('getObjectKeys', () => {
  test('gets own properties', () => {
    const result = getObjectKeys(testObj);

    expect(result).toMatchSnapshot();
  });

  test('gets enumerable properties', () => {
    const result = getObjectKeys(testObj, true);

    expect(result).toMatchSnapshot();
  });
});

describe('isObject', () => {
  test('true of object literal', () => {
    expect(isObject({})).toBe(true);
  });

  test('false for array', () => {
    expect(isObject([])).toBe(false);
  });

  test('false for function', () => {
    expect(isObject(Function.prototype)).toBe(false);
  });
});

describe('isPropDescriptor', () => {
  test('false when not an object', () => {
    const result = isPropDescriptor(null);

    expect(result).toBe(false);
  });

  test('false without a descriptor property', () => {
    const result = isPropDescriptor({ a: 1 });

    expect(result).toBe(false);
  });

  test('true with value and another descriptor prop', () => {
    const result = isPropDescriptor({
      configurable: true,
      value: true
    });

    expect(result).toBe(true);
  });

  test('true with get and a descriptor prop', () => {
    const result = isPropDescriptor({
      get() {
        return true;
      },
      enumerable: true
    });

    expect(result).toBe(true);
  });

  test('true with set and a descriptor prop', () => {
    const result = isPropDescriptor({
      set() {
        this.test = true;
      },
      enumerable: true
    });

    expect(result).toBe(true);
  });

  test('true with only get and set', () => {
    const result = isPropDescriptor({
      get() {
        return true;
      },
      set() {
        this.test = true;
      }
    });

    expect(result).toBe(true);
  });
});
