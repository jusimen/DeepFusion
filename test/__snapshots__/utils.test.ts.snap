// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`countOccurrences generates map 1`] = `
{
  "a": 1,
  "b": 2,
  "c": 3,
  "d": 2,
  "e": 1,
}
`;

exports[`getInAll gets universal values in multiple arrays 1`] = `
[
  "c",
]
`;

exports[`getInAll gets universal values in two arrays 1`] = `
[
  "b",
  "c",
]
`;

exports[`getInMultiple gets common values found in multiple arrays 1`] = `
[
  "b",
  "c",
  "d",
]
`;

exports[`getInMultiple gets common values in two arrays 1`] = `
[
  "b",
  "c",
]
`;

exports[`getNotInAll gets values not found in three arrays 1`] = `
[
  "a",
  "b",
  "d",
  "e",
]
`;

exports[`getNotInAll gets values not found in two arrays 1`] = `
[
  "a",
  "d",
]
`;

exports[`getNotInMultiple gets values found in one of multiple arrays 1`] = `
[
  "a",
  "e",
]
`;

exports[`getNotInMultiple gets values found in one of two arrays 1`] = `
[
  "a",
  "d",
]
`;

exports[`getObjectKeys gets enumerable properties 1`] = `
[
  "firstName",
  "lastName",
  "fullName",
  "enumerableTrue",
]
`;

exports[`getObjectKeys gets own properties 1`] = `
[
  "firstName",
  "lastName",
  "fullName",
]
`;
